Fix filling parts columns from metadata (when columns.txt does not exists)

azat opened the pull request:
The getFileNameForColumn() inside IMergeTreeDataPart::loadColumns()
requries serializations that will be filled only at the end of
loadColumns() (setColumns()):

    (lldb) bt
    * thread 718, name = 'ActiveParts', stop reason = breakpoint 1.1
      * frame 0: 0x00000000111e50dc clickhouse`DB::IMergeTreeDataPart::getSerialization(this=0x00007ffd18c39918, column_name="key") const + 124 at IMergeTreeDataPart.cpp:508
        frame 1: 0x0000000011340f57 clickhouse`DB::MergeTreeDataPartWide::getFileNameForColumn(this=0x00007ffd18c39918, column=<unavailable>) const + 55 at MergeTreeDataPartWide.cpp:301
        frame 2: 0x00000000111e9a2b clickhouse`DB::IMergeTreeDataPart::loadColumns(this=0x00007ffd18c39918, require=false) + 1067 at IMergeTreeDataPart.cpp:1577
        frame 3: 0x00000000111e8d52 clickhouse`DB::IMergeTreeDataPart::loadColumnsChecksumsIndexes(this=0x00007ffd18c39918, require_columns_checksums=<unavailable>, check_consistency=true) + 82 at IMergeTreeDataPart.cpp:713
        frame 4: 0x0000000011284c10 clickhouse`DB::MergeTreeData::loadDataPart(this=0x00007ffcf9242e40, part_info=0x00007ffcf937c298, part_name="2024", part_disk_ptr=std::__1::shared_ptr<DB::IDisk>::element_type @ 0x00007ffcf925af18 strong=11 weak=2, to_state=Active, part_loading_mutex=0x00007ffef97efeb0) + 3152 at MergeTreeData.cpp:1435

So let's add fallback into the getFileNameForColumn() in case of
serializations was not filled yet.

Note, that it is OK to add this check as a generic, since after
loadColumns() serializations cannot be empty, since loadColumns() does
not allows this (there is only one more caller of
getFileNameForColumn() - loadIndexGranularity())

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix filling parts columns from metadata (when columns.txt does not exists)

Cc: @CurtizJ

azat committed:
Fix filling parts columns from metadata (when columns.txt does not exists)

The getFileNameForColumn() inside IMergeTreeDataPart::loadColumns()
requries serializations that will be filled only at the end of
loadColumns() (setColumns()):

    (lldb) bt
    * thread 718, name = 'ActiveParts', stop reason = breakpoint 1.1
      * frame 0: 0x00000000111e50dc clickhouse`DB::IMergeTreeDataPart::getSerialization(this=0x00007ffd18c39918, column_name="key") const + 124 at IMergeTreeDataPart.cpp:508
        frame 1: 0x0000000011340f57 clickhouse`DB::MergeTreeDataPartWide::getFileNameForColumn(this=0x00007ffd18c39918, column=<unavailable>) const + 55 at MergeTreeDataPartWide.cpp:301
        frame 2: 0x00000000111e9a2b clickhouse`DB::IMergeTreeDataPart::loadColumns(this=0x00007ffd18c39918, require=false) + 1067 at IMergeTreeDataPart.cpp:1577
        frame 3: 0x00000000111e8d52 clickhouse`DB::IMergeTreeDataPart::loadColumnsChecksumsIndexes(this=0x00007ffd18c39918, require_columns_checksums=<unavailable>, check_consistency=true) + 82 at IMergeTreeDataPart.cpp:713
        frame 4: 0x0000000011284c10 clickhouse`DB::MergeTreeData::loadDataPart(this=0x00007ffcf9242e40, part_info=0x00007ffcf937c298, part_name="2024", part_disk_ptr=std::__1::shared_ptr<DB::IDisk>::element_type @ 0x00007ffcf925af18 strong=11 weak=2, to_state=Active, part_loading_mutex=0x00007ffef97efeb0) + 3152 at MergeTreeData.cpp:1435

So let's add fallback into the getFileNameForColumn() in case of
serializations was not filled yet.

Note, that it is OK to add this check as a generic, since after
loadColumns() serializations cannot be empty, since loadColumns() does
not allows this (there is only one more caller of
getFileNameForColumn() - loadIndexGranularity())

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix "No columns in part" check

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #64757 from azat/ClickHouse:load-part-without-columns.txt -->
*This is an automated comment for commit 13cc3280c0a99e2432af3c1a8a436637dfe22655 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64757/13cc3280c0a99e2432af3c1a8a436637dfe22655/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

azat committed:
Use primary.cidx instead of columns.txt to break parts

After fixes for columns.txt the test_max_suspicious_broken_parts starts
to fail

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>Integration tests (aarch64) [5/6] — fail: 1, passed: 582

- https://github.com/ClickHouse/ClickHouse/issues/64451

>Stateless tests (aarch64) — fail: 1, passed: 6616, skipped: 26

- `02440_mutations_finalization` is flaky

azat commented:
@SmitaRKulkarni should it be rebased or what?

SmitaRKulkarni commented:
@azat : Sorry not clear why do you feel the need to rebase ?

azat commented:
@SmitaRKulkarni Because it hasn't been applied yet

SmitaRKulkarni commented:
> @SmitaRKulkarni Because it hasn't been applied yet

I dont think we need rebase, as you mentioned the test fails are not related. I will try to merge it.
